### PR TITLE
add filters before loss function computation

### PR DESCRIPTION
**Pull request by AldoGl**
## Proposed changes

I add the possibility of adding a list of callables to act on each dimension of the simulated time series before the loss is evaluated. I also add a specific HP filter in the utils/time_series.py module. Finally I add a test for the new feature

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../blob/master/CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by AldoGl** [comment]
Hi @marcofavoritobi, following your suggestion I added the possibility of specifying a mixed list of the kind `[function, None, function, function]` at the list of coordinate filters in the loss computation, if `None` is selected no filter is applied to the coordinate

File: black_it/loss_functions/base.py [hunk 1]
```diff
@@ -66,12 +76,39 @@ def compute_loss(
...
+        else:
+            nb_coordinate_filters = len(self.coordinate_filters)
+            assert_(
+                nb_coordinate_filters == num_coords,
+                (
+                    "the length of coordinate_filters should be equal "
+                    f"to the number of coordinates, got {nb_coordinate_filters} and {num_coords}"
+                ),
+                exc_cls=ValueError,
+            )
```
Comment (marcofavoritobi) [inline black_it/loss_functions/base.py:91]: I guess this validation could be moved into the constructor; otherwise an error would occur only when `compute_loss` is called, which could surprise the user a bit

File: black_it/loss_functions/base.py [hunk 2]
```diff
@@ -66,12 +76,39 @@ def compute_loss(
...
-                self.compute_loss_1d(sim_data_ensemble[:, :, i], real_data[:, i])
-                * weights[i]
-            )
+            if filters[i] is None:
+                filtered_data = np.array(
+                    [
+                        sim_data_ensemble[j, :, i]
+                        for j in range(sim_data_ensemble.shape[0])
+                    ]
+                )
```
Comment (marcofavoritobi) [inline black_it/loss_functions/base.py:102]:
```suggestion
                filtered_data = sim_data_ensemble[:, :, i]
```

File: black_it/loss_functions/base.py (hunk 1 above)
Comment (AldoGl) [reply black_it/loss_functions/base.py:91]: We discussed this already. Unfortunately it is not possible since the BaseLoss constructor does not know the number of coordinates when it's called

File: black_it/loss_functions/base.py [hunk 3]
```diff
@@ -66,12 +76,39 @@ def compute_loss(
...
+            )
+            filters = self.coordinate_filters
+
         loss = 0
         for i in range(num_coords):
-            loss += (
-                self.compute_loss_1d(sim_data_ensemble[:, :, i], real_data[:, i])
-                * weights[i]
-            )
+            if filters[i] is None:
```
Comment (marcofavoritobi) [inline black_it/loss_functions/base.py:96]: An idea to simplify the code is to move this check in the constructor, pre-process the `coordinate_filters` argument so to have either user-defined filters or the identity functions.

File: black_it/loss_functions/base.py (hunk 2 above)
Comment (AldoGl) [reply black_it/loss_functions/base.py:102]: Yes it is, thanks, I will fix it

File: black_it/utils/time_series.py [hunk 4]
```diff
@@ -81,3 +84,41 @@ def get_mom_ts_1d(time_series: NDArray[np.float64]) -> NDArray[np.float64]:
     np.nan_to_num(avg_vec_mom, False)
 
     return avg_vec_mom
+
+
+def hp_filter(
+    time_series: NDArray[np.float64], lamb: float = 1600
+) -> Tuple[NDArray[np.float64], NDArray[np.float64]]:
+    """
+    Apply the HP filter to a time series.
```
Comment (marcofavoritobi) [inline black_it/utils/time_series.py:93]: Might be a good idea to put a bibliographic reference here?

File: black_it/loss_functions/base.py (hunk 3 above)
Comment (AldoGl) [reply black_it/loss_functions/base.py:96]: Once again, this is not possible since the constructor does not know the number of coordinates unfortunately

File: black_it/utils/time_series.py (hunk 4 above)
Comment (AldoGl) [reply black_it/utils/time_series.py:93]: Yes, thanks

File: black_it/loss_functions/base.py (hunk 1 above)
Comment (marcofavoritobi) [reply black_it/loss_functions/base.py:91]: right, but a sanity check with `coordinate_weights` could be useful

File: black_it/loss_functions/base.py [hunk 5]
```diff
@@ -17,25 +17,35 @@
...
+
 
 class BaseLoss(ABC):
     """BaseLoss interface."""
 
-    def __init__(self, coordinate_weights: Optional[NDArray] = None):
+    def __init__(
+        self,
+        coordinate_weights: Optional[NDArray] = None,
+        coordinate_filters: Optional[List[Optional[Callable]]] = None,
```
Comment (marcofavoritobi) [inline black_it/loss_functions/base.py:37]: I guess we could have three behaviours of this parameter:

1. None -> the identity function
2. callable -> a filter that will be applied to every dimension
3. list of callables/none -> a filter specified for each dimension

1 and 3 are in place; if useful, we could add 2 (*even after the PR is merged*)

File: black_it/loss_functions/base.py (hunk 3 above)
Comment (marcofavoritobi) [reply black_it/loss_functions/base.py:96]: I think this part is possible because in case `filters[i]` is None, we can assign a lambda function like `filters[i] = lambda x: x`, and then just use it rather than checking `if filters[i] is None`

File: black_it/loss_functions/base.py (hunk 5 above)
Comment (AldoGl) [reply black_it/loss_functions/base.py:37]: Thanks this might be a good idea for a future PR if we find it to be potentially useful. For the moment I would not implement it though

Comment (marcofavoritobi) [reply black_it/loss_functions/base.py:37]: Another idea we talked about is to use dictionaries rather than lists.

For example, say we have 10 dimensions, but we want to apply a filter for the 3rd, the 5th and the 7th dimensions. With the current interface we had to do something like:

```python
[None] * 2 + [filter_for_3rd] + [None] + [filter_for_5th] + [None] + [filter_for_7th]  + [None] * 3
```

Instead, if we used dictionaries with dimension indexes as keys, we could do:
```python
{
  2 : filter_for_3rd # (or 3?)
  4 : filter_for_5th # (or 5?)
  6 : filter_for_7th # (or 7?)
}
```

**Comment by AldoGl** [comment]
Thank you @muxator for your comment on the correction of the `hp_filter` docstring. I implemented that too

File: black_it/utils/time_series.py [hunk 6]
```diff
@@ -81,3 +84,45 @@ def get_mom_ts_1d(time_series: NDArray[np.float64]) -> NDArray[np.float64]:
...
+    Args:
+        time_series: the one dimensional series to be filtered
+        lamb: the Hodrick-Prescott smoothing parameter, a value of 1600 is suggested for quarterly data
+
+    Returns:
+        the cycle and the trend of the HP decomposition of the time series
+
+    References:
+        Hodrick, R.J, and E. C. Prescott. 1980. "Postwar U.S. Business Cycles: An Empirical Investigation."
+        Carnegie Mellon University discussion paper no. 451`.
```
Comment (marcofavoritobi) [inline black_it/utils/time_series.py:104]:
```suggestion
        Carnegie Mellon University discussion paper no. 451.
```

:slightly_smiling_face:

Comment (AldoGl) [reply black_it/utils/time_series.py:104]: Well spotted!!

**Comment by AldoGl** [comment]
Thank you @marcofavoritobi, I am closing this one
